chain/chain/src/doomslug.rs

```diff
@@ -0,0 +1,1001 @@
+        self.witness.entry(approval.account_id.clone()).or_insert_with(|| {
+            increment_approved_stake = true;
+            if approval.is_endorsement {
+                increment_endorsed_stake = true;
+            }
+            approval.clone()
+        });
+
+        if increment_approved_stake {
+            self.approved_stake +=
+                self.account_id_to_stake.get(&approval.account_id).map_or(0, |x| *x);
+        }
+
+        if increment_endorsed_stake {
+            self.endorsed_stake +=
+                self.account_id_to_stake.get(&approval.account_id).map_or(0, |x| *x);
+        }
+
+        // We call to `get_block_production_readiness` here so that if the number of approvals crossed
+        // the threshold, the timer for block production starts.
+        self.get_block_production_readiness(now)
+    }
+
+    /// Withdraws an approval. This happens if a newer approval for the same `target_height` comes
+    /// from the same account. Removes the approval from the `witness` and updates approved and
+    /// endorsed stakes.
+    fn withdraw_approval(&mut self, account_id: &AccountId) {
+        let approval = match self.witness.remove(account_id) {
+            None => return,
+            Some(approval) => approval,
+        };
+
+        self.approved_stake -= self.account_id_to_stake.get(&approval.account_id).map_or(0, |x| *x);
+
+        if approval.is_endorsement {
+            self.endorsed_stake -=
+                self.account_id_to_stake.get(&approval.account_id).map_or(0, |x| *x);
+        }
+    }
+
+    /// Returns whether the block is ready to be produced without waiting for any extra time,
+    /// has crossed the doomslug threshold (and thus can be produced given enough time passed),
+    /// or is not ready to be produced at all.
+    /// This method only returns `ReadyToProduce` if the block has 2/3 approvals and 1/2 endorsements.
+    /// The production readiness due to enough time passing since crossing the threshold should be
+    /// handled by the caller.
+    ///
+    /// # Arguments
+    /// * now - the current timestamp
+    ///
+    /// # Returns
+    /// `None` if the block doesn't have enough approvals yet to cross the doomslug threshold
+    /// `PassedThreshold` if the block has enough approvals to pass the threshold, but can't be
+    ///     produced bypassing the timeout
+    /// `ReadyToProduce` if the block can be produced bypassing the timeout (i.e. has 2/3+ of
+    ///     approvals and 1/2+ of endorsements)
+    fn get_block_production_readiness(&mut self, now: Instant) -> DoomslugBlockProductionReadiness {
+        if self.approved_stake > self.total_stake * 2 / 3
+            && self.endorsed_stake > self.total_stake / 2
+        {
+            if self.time_passed_threshold == None {
+                self.time_passed_threshold = Some(now);
+            }
+            DoomslugBlockProductionReadiness::ReadyToProduce(self.time_passed_threshold.unwrap())
```

Comment:
I suggest to avoid `.unwrap()`, .e.g.:

```rust
let time_passed_threshold = if let Some(threshold) = self.time_passed_threshold {
    threshold
} else {
    self.time_passed_threshold = Some(now);
    now
};
DoomslugBlockProductionReadiness::ReadyToProduce(time_passed_threshold)
```